### ISSUE TITLE
[Feature] Built-in adapter for image comparison library 

### ISSUE THREAD

**Issue by dmtrKovalenko** [microsoft/playwright#28578]
We are doing a lot of visual regression tests during out CI and I noticed that when tests are running with and without snapshots there is a significant time difference for making and comparing screenshots. 

So I build and now use https://github.com/dmtrKovalenko/odiff for my playwright tests. The only addition I'd like to have is the ability to specify the comparison library in playwright instead of completely rewriting the logic of comparing screenshots. 

Example

```ts
// playwright.config.ts
compareImages: (originalPath: string, actualPath: string, diffPath: string, width, height, options) => { 
   // here is all the logic related to comparing screenshots and outputting the diff
}
```

Which is by default equals to 
https://github.com/microsoft/playwright/blob/main/packages/playwright-core/src/third_party/pixelmatch.js

I'd be happy to make a PR if there will be a green light

**Comment by gskierk** [comment]
How would you like to pass on information about the library? In the configuration file? In the `.toHaveScreenshot()` method? So we would need to define an interface for adapters first.

**Comment by dmtrKovalenko** [comment]
Yeah just make an interface and pass an adapter in the configuration

**Comment by dmtrKovalenko** [comment]
@pavelfeldman what is wrong with this issue? It even has a lot of likes

**Comment by pavelfeldman** [comment]
Nothing wrong with the issue, I can bring it back into the triaging state. Extensibility would imply committing to the new public API, documentation, tests, etc. Let's start with agreeing on the shape of the API and then we can proceed with the implementation.

**Comment by dmtrKovalenko** [comment]
The API I need is anything just being accessible or exposed by playwright. Ideally I need playwright to store the screenshot, give me the **paths** to the files and then I can correctly (not from javascript) handle io, comparison and saving the diff

the ideal implementation though will likely take you forever to implement, so what I am asking for is just exposed your current API that I can implement myself to compare the images and return the result back. The proposed change I made does not require anything from playwright just exposing _comparator option not just as a string but also as a function that is replacing the existing comparator.

**Comment by pavelfeldman** [comment]
There is no "current api", so whatever you want to do will be our first api...

**Comment by dmtrKovalenko** [comment]
There is an existing api for comparator that takes 2 buffers and output buffer of difference and error message or undefined. 

I can use that or it would be better to have something like this:


```
customImagePathComparator: (currentPath: string, newPath: string, options: any) => { matched: false, message: string, diffCount: number } | { matched: true }

customImageBufferComparator: (mimeType: string, currentPath: Uint8Array, newPath: UInt8Array, options: any) => { matched: false, message: string, diffCount: number } | { matched: true }
```

Where the options would by dynamic to support any options passed to playwright by the user. Essentially threshold per screenshot. 

For the end user configuration should be defined once, so having it at the config level of screenshot matching is perfect. 

I will provide a direct implementation from odiff so the config will look like

```
import odiffComparator from "odiff-bin/playwright"

...
toHaveScreenshot: {
  comparator: odiffComparator
}
```

**Comment by pavelfeldman** [comment]
> There is an existing api for comparator 

Could you point to the Playwright docs where this API is documented?

**Comment by dmtrKovalenko** [comment]
It is not documented it is your internal API I'm happy to implement drop in replacement for if you will allow me

**Comment by pavelfeldman** [comment]
In this case I already replied here: https://github.com/microsoft/playwright/issues/28578#issuecomment-3439086834

**Comment by dmtrKovalenko** [comment]
Okay, then what do you think about the API I proposed? Can I implement it?

**Comment by pavelfeldman** [comment]
This does not sound like this is converging. I'll close it as out of scope, sorry.

**Comment by dmtrKovalenko** [comment]
* Provide a feature request people need
* People do +1s
* Maintainers absolutely ignore the issue 
* You implement the solution in 25 lines of code not requiring maintenance from playwright at all 
* Unkooth maintainer closes your PR because he can't see an issue in the description
* You pointing him to the issue
* HE CAN'T UNDERSTAND
* your provide rationale and a real API design YOU CAN IMPLEMENT 
* he answer "this doesn't sound like this is converging* 

<img width="600" height="337" alt="Image" src="https://github.com/user-attachments/assets/72fab075-e9ab-4f44-9a10-d3cc0b54d85f" />

**Comment by Nightquaker** [comment]
KEKW
Average corpo OSS maintainer behavior.

**Comment by hvent90** [comment]
Hi chiming in - I would personally benefit from this being added.

**Comment by 4g2725qbsv-png** [comment]
hey so uhm. are we fucking retarded

**Comment by alexhraber** [comment]
> This does not sound like this is converging. I'll close it as out of scope, sorry.

This is exactly why capitalism doesn't mesh with higher-minded ideas such as "open source".

**Comment by alexhraber** [comment]
> > This does not sound like this is converging. I'll close it as out of scope, sorry.
> 
> This is exactly why capitalism doesn't mesh with higher-minded ideas such as "open source".

If @microsoftopensource is hiring, I'll take this guy's role for 30% his salary.

**Comment by stanislavkozlovski** [comment]
@microsoftopensource wtf is this behavior, you call this open-source?

**Comment by Eclips4** [comment]
It’s pretty clear that there’s no agreement on the API. Instead of having a constructive discussion, you’re turning it into a nervous back-and-forth. Then, for some reason, you share it with a bunch of people who haven’t even read the maintainer comments and add nothing to the conversation except unnecessary rage-bait.

FYI @dmtrKovalenko, that’s not how opensource works.

**Comment by Lyamc** [comment]
@Eclips4 I'm sorry, what?

Maintainer: You can submit an API and it'll be the first API
Submitter: Here you go, what do you think of it?
Maintainer: Closing the PR because reasons™

The maintainer closed the PR without saying WHAT isn't converging, ending any further discussion.

**Comment by Eclips4** [comment]
@Lyamc Yeah, I don’t deny that there’s something strange on both sides :)
I’m talking about bringing in a lot of people here, which isn’t helping in this case, it just creates more hate and unnecessary messages (this message is one of them). I’m saying this is a very odd approach to solving a communication problem.

**Comment by uhx** [comment]
> @Eclips4 I'm sorry, what?
> 
> Maintainer: You can submit an API and it'll be the first API
> Submitter: Here you go, what do you think of it?
> Maintainer: Closing the PR because reasons™
> 
> The maintainer closed the PR without saying WHAT isn't converging, ending any further discussion. 

You skipped this part:

Maintainer:

> Let's start with agreeing on the shape of the API and then we can proceed with the implementation.

Submitter:

> the ideal implementation though will likely take you forever to implement, so what I am asking for is just exposed your current API

[looks like refusal to follow the proposed process]

"open-source" doesn't mean you can push whatever you want. If this API is internal, it should not be used or patched to fit user-specific needs. 

Making it public requires additional steps; thats normal.

Since it open-source, you're free to create your own fork.

**Comment by Lyamc** [comment]
@uhx 
> [looks like refusal to follow the proposed process]

It "looks like" that because the maintainer didn't actually say the reason, which means we're now all free to assume whatever we want. Do you not understand that my criticism is in the failure to communicate? That's on the maintainer, not the submitter.

**Comment by golergka** [comment]
> This does not sound like this is converging. 

@pavelfeldman what does "converging" mean in this context? How can outside contributors to this repo make sure that their contributions are "converging" so that they bring more value to this project and the community?

**Comment by Lewiscowles1986** [comment]
This is the worst, most entitled behaviour. He said no, and pointed out, he would need to understand, agree with, and maintain your hot-takes. You admitted you were using an internal, undocumented API. It's OpenSource, you can fork it if you really need this. Nobody is blocking you guys.

One person has two PR's, one with just no chill at-all, all caps due to the other being closed.

I Am on the side of some OSS maintainers being a bit... different to how I would like, but the one you are targetting has contributed a lot more to this, than anyone commenting. They are bearing the burden of maintenance. This is not a department store you can loot; so please, start maintaining a fork, if it is this important.

**Comment by io-june** [comment]
valid crashout tbh

**Comment by gamer191** [comment]
Let me rephrase @dmtrKovalenko's comments:

|@dmtrKovalenko's comment|comment meaning|
|-|-|
|the ideal implementation though will likely take you forever to implement|I would like to implement a poor quality implementation, because I won't have to spend time dealing with the technical debt|
|The proposed change I made does not require anything from playwright just exposing _comparator option not just as a string but also as a function that is replacing the existing comparator.|I don't think that making an internal API public requires any increased maintenance effort. I don't understand why every programming language has internal functions, so I will assume that there's no valid reason (btw, the reason is that APIs shouldn't be exposed unless you're willing to commit to maintaining them|
|Maintainers absolutely ignore the issue|There's 536 open issues, why isn't my issue being prioritised😠|
|The API I need is anything just being accessible or exposed by playwright|I don't care about code quality so you shouldn't either|


I've read through the context (this issue and the associated PRs), so I hope this is an accurate description of @dmtrKovalenko's tantrum. I generally wouldn't comment this on a repo I'm unfamiliar with, but @dmtrKovalenko invited everyone to comment on this issue, and I think its fair to say that I've read more of the context than his average twitter fan reacting with a thumbs down and an insulting comment

**Comment by hvent90** [comment]
> Let me rephrase [@dmtrKovalenko](https://github.com/dmtrKovalenko)'s comments:
> 
> [@dmtrKovalenko](https://github.com/dmtrKovalenko)'s comment	comment meaning
> the ideal implementation though will likely take you forever to implement	I would like to implement a poor quality implementation, because I won't have to spend time dealing with the technical debt
> The proposed change I made does not require anything from playwright just exposing _comparator option not just as a string but also as a function that is replacing the existing comparator.	I don't think that making an internal API public requires any increased maintenance effort. I don't understand why every programming language has internal functions, so I will assume that there's no valid reason (btw, the reason is that APIs shouldn't be exposed unless you're willing to commit to maintaining them
> Maintainers absolutely ignore the issue	There's 536 open issues, why isn't my issue being prioritised😠
> The API I need is anything just being accessible or exposed by playwright	I don't care about code quality so you shouldn't either
> I've read through the context (this issue and the associated PRs), so I hope this is an accurate description of [@dmtrKovalenko](https://github.com/dmtrKovalenko)'s tantrum. I generally wouldn't comment this on a repo I'm unfamiliar with, but [@dmtrKovalenko](https://github.com/dmtrKovalenko) invited everyone to comment on this issue, and I think its fair to say that I've read more of the context than his average twitter fan reacting with a thumbs down and an insulting comment

Not digging the condescending attitude and belittling everyone whose takeaway is different than yours. It’s safe to say folks’ reaction to this is coming from a valid place and there’s something to learn here.

**Comment by Lyamc** [comment]
@gamer191 the maintainer didn't even read the first two words of the PR that was closed. You clearly didn't either, and therefore have missed the point entirely.

**Comment by pavelfeldman** [comment]
Your angry comments about a lousy corporate open source maintainer made me update my GitHub bio to flex and gain a little respect 😄. Let me respond since it caught a lot of attention.

> @pavelfeldman what does "converging" mean in this context? How can outside contributors to this repo make sure that their contributions are "converging" so that they bring more value to this project and the community?

### Process

I stated the requirements to the new APIs here: https://github.com/microsoft/playwright/issues/28578#issuecomment-3439086834:

<img width="867" height="182" alt="Image" src="https://github.com/user-attachments/assets/a7416702-162a-4ae8-9e26-861ac696a63b" />

I'm not sure why the wave of the downvotes to be honest, it all sounds pretty innocent.

### Converging

"Converging" means we are making progress, progressing towards a solution:
- API shape is being presented and elaborated
- Design comments are being addressed
- Contributor understands long-term API committment
- Contributor demonstrates readiness to produce necessary changes including docs and tests
 
That's how it usually works, but unfortunately none of this happened in this issue. Contributor was only willing to expose an implementation detail of Playwright to hook the library they authored into it.

### Call

When contributor [edits the file](https://github.com/microsoft/playwright/pull/37983/files#diff-d7fe4faa545d7a7622110cfd9d4274d9456fc68a454537fe9423f94fedef1c1dR19) under the line "do not edit manually", you make conclusions, it means that there is long road of reviews ahead. It is not a show stopper, but it is a signal.

<img width="685" height="165" alt="Image" src="https://github.com/user-attachments/assets/c859c1fe-a3bb-40d9-af53-3b0e129f0f02" />

Now we have to make a judgement call:
- The feature request has been filed by the contributor
- Only collected 10 upvotes over 2 years (very low number on Playwright scale)
- Contributor is reluctant to elaborate on the API or even admit the fact that it will be an API
- The first take on the PR is not great and is likely to consume a lot of time.

---

I hope it helps. Btw, I think that the idea is worth pursuing, so if someone wants to follow up, please file an issue, document the API with the intended behavior and types, commit to the implementation. Let's turn all this energy into the code and features!

**Comment by pavelfeldman** [comment]
@Lyamc

> @gamer191 the maintainer didn't even read the first two words of the PR that was closed. You clearly didn't either, and therefore have missed the point entirely.

This was my mistake - instead of suggesting to file a new issue, I should have brought this issue into the triaging state. Which I did the moment I realized it exists. We always close the PRs that have not been discussed in the Issue though. In this case PR came for an issue that is 2 y.o., which caught me off guard. I hope it makes sense.

**Comment by golergka** [comment]
@pavelfeldman thank you for a polite and detailed answer. I have to admit that after reading these replies I have changed my view of this discussion.
